keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ImageResolver.kt

```diff
@@ -0,0 +1,89 @@
+package com.netflix.spinnaker.keel.ec2.resource
+
+import com.netflix.spinnaker.keel.api.NoImageFound
+import com.netflix.spinnaker.keel.api.NoImageFoundForRegion
+import com.netflix.spinnaker.keel.api.NoImageSatisfiesConstraints
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.UnsupportedStrategy
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.api.ec2.image.ArtifactImageProvider
+import com.netflix.spinnaker.keel.api.ec2.image.IdImageProvider
+import com.netflix.spinnaker.keel.api.ec2.image.ImageProvider
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.clouddriver.model.NamedImageComparator
+import com.netflix.spinnaker.keel.clouddriver.model.appVersion
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import org.slf4j.LoggerFactory
+
+class ImageResolver(
+  private val dynamicConfigService: DynamicConfigService,
+  private val cloudDriverService: CloudDriverService,
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  constraintEvaluators: List<ConstraintEvaluator<*>>
+) {
+
+  private val constraintEvaluators = constraintEvaluators
+    .associateBy { it.constraintType }
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  // TODO: we may (probably will) want to make the parameter type more generic
+  suspend fun resolveImageId(resource: Resource<ClusterSpec>): String {
+    when (val imageProvider = resource.spec.launchConfiguration.imageProvider) {
+      is IdImageProvider -> {
+        return imageProvider.imageId
+      }
+      is ArtifactImageProvider -> {
+        val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(resource)
+        val environment = deliveryConfigRepository.environmentFor(resource)
+        val artifact = imageProvider.deliveryArtifact
+        val artifactName = artifact.name
+        val account = dynamicConfigService.getConfig("images.default-account", "test")
+        val namedImage = cloudDriverService
+          .namedImages(artifactName, account)
+          .sortedWith(NamedImageComparator)
+          .let { images ->
+            if (images.isEmpty()) {
+              throw NoImageFound(artifactName)
+            } else if (deliveryConfig == null || environment == null || environment.constraints.isEmpty()) {
+              images.last()
+            } else {
+              images.lastOrNull { image ->
+                constraintEvaluators.all {
+                  it.value.canPromote(artifact, image.appVersion, deliveryConfig, environment.name)
+                }
+              }
```

Comment:
… otherwise find the newest image that satisfies all constraints on the environment